File: claude-code/seeds/code-feature.js
// Seed: code-feature — implement a feature across files: plan, write per-file, self-review.
// Claude Code Workflow edition. args: { spec, files: [{path, content}] }
export const meta = {
  name: 'code-feature',
  description: 'Implement a feature: plan the minimal edits, generate per-file changes, self-review each.',
  phases: [{ title: 'Plan' }, { title: 'Implement' }, { title: 'Review' }],
}

const spec = (args && args.spec) || ''
const files = (args && args.files) || []
const byPath = Object.fromEntries(files.map((f) => [f.path, f.content]))

const PLAN_SCHEMA = {
  type: 'object', additionalProperties: false, required: ['steps'],
  properties: { steps: { type: 'array', items: {
    type: 'object', additionalProperties: false, required: ['path', 'change'],
    properties: { path: { type: 'string' }, change: { type: 'string' } } } } },
}
const NEW_SCHEMA = {
  type: 'object', additionalProperties: false, required: ['content', 'note'],
  properties: { content: { type: 'string' }, note: { type: 'string' } },
}
const REVIEW_SCHEMA = {
  type: 'object', additionalProperties: false, required: ['approved', 'issues'],
  properties: { approved: { type: 'boolean' }, issues: { type: 'string' } },
}

phase('Plan')
const plan = await agent(
  `Plan the MINIMAL edits to implement this feature:\n${spec}\n\nExisting files:\n` +
  files.map((f) => `- ${f.path}`).join('\n'),
  { label: 'plan', phase: 'Plan', model: 'opus', schema: PLAN_SCHEMA })
const steps = (plan && plan.steps) || []

phase('Implement')
const edits = (await parallel(steps.map((step) => () =>
  agent('Apply this change and return the FULL new file content.\n' +
        `CHANGE: ${step.change}\nFEATURE SPEC: ${spec}\n\nCURRENT (${step.path}):\n${byPath[step.path] || ''}`,
        { label: `impl:${step.path}`, phase: 'Implement', model: 'sonnet', schema: NEW_SCHEMA })
    .then((r) => ({ path: step.path, content: (r && r.content) || '', note: (r && r.note) || '' }))))).filter(Boolean)

phase('Review')
const reviewed = (await parallel(edits.map((e) => () =>
  agent('Review this edit against the spec. Flag bugs/omissions; approve only if correct.\n' +
        `SPEC: ${spec}\nFILE ${e.path}:\n${e.content}`,
        { label: `review:${e.path}`, phase: 'Review', model: 'opus', schema: REVIEW_SCHEMA })
    .then((v) => ({ ...e, approved: !!(v && v.approved), issues: (v && v.issues) || '' }))))).filter(Boolean)

const approved = reviewed.filter((r) => r.approved).length
log(`${approved}/${reviewed.length} edits approved`)
return { edits: reviewed, approved, total: reviewed.length }

File: seeds/code-feature.py
"""Seed: code-feature — implement a feature across files: plan, write per-file, self-review.

args: {"spec": "...", "files": [{"path","content"}, ...]}
Coding seeds pair well with the optional execution grader (graders/pytest_grader.py)
in run-all mode: swap the review agent for real test runs to get an objective reward.
"""

META = {
    "name": "code-feature",
    "description": "Implement a feature: plan, generate per-file changes, self-review.",
    "phases": ["Plan", "Implement", "Review"],
}

PLAN_SCHEMA = {
    "type": "object", "additionalProperties": False,
    "properties": {"steps": {"type": "array", "items": {
        "type": "object", "additionalProperties": False,
        "properties": {"path": {"type": "string"}, "change": {"type": "string"}},
        "required": ["path", "change"]}}},
    "required": ["steps"],
}
NEW_SCHEMA = {
    "type": "object", "additionalProperties": False,
    "properties": {"content": {"type": "string"}, "note": {"type": "string"}},
    "required": ["content", "note"],
}
REVIEW_SCHEMA = {
    "type": "object", "additionalProperties": False,
    "properties": {"approved": {"type": "boolean"}, "issues": {"type": "string"}},
    "required": ["approved", "issues"],
}


async def run(agent, parallel, pipeline, log, phase, args):
    a = args or {}
    spec = a.get("spec", "")
    files = a.get("files", [])                      # [{"path","content"}]
    by_path = {f["path"]: f["content"] for f in files}

    phase("Plan")
    plan = await agent(
        f"Plan the MINIMAL edits to implement this feature:\n{spec}\n\nExisting files:\n"
        + "\n".join(f"- {f['path']}" for f in files),
        label="plan", phase="Plan", model="opus", schema=PLAN_SCHEMA)
    steps = (plan or {}).get("steps", [])

    phase("Implement")
    async def implement(step):
        cur = by_path.get(step["path"], "")
        r = await agent(
            f"Apply this change and return the FULL new file content.\n"
            f"CHANGE: {step['change']}\nFEATURE SPEC: {spec}\n\n"
            f"CURRENT ({step['path']}):\n{cur}",
            label=f"impl:{step['path']}", phase="Implement", model="sonnet", schema=NEW_SCHEMA)
        return {"path": step["path"], "content": (r or {}).get("content", ""),
                "note": (r or {}).get("note", "")}
    edits = [e for e in await parallel([(lambda s=s: implement(s)) for s in steps]) if e]

    phase("Review")
    async def review(e):
        v = await agent(
            "Review this edit against the spec. Flag bugs/omissions; approve only if correct.\n"
            f"SPEC: {spec}\nFILE {e['path']}:\n{e['content']}",
            label=f"review:{e['path']}", phase="Review", model="opus", schema=REVIEW_SCHEMA)
        return {**e, "approved": bool(v and v.get("approved")), "issues": (v or {}).get("issues", "")}
    reviewed = [r for r in await parallel([(lambda e=e: review(e)) for e in edits]) if r]

    approved = sum(1 for r in reviewed if r["approved"])
    log(f"{approved}/{len(reviewed)} edits approved")
    return {"edits": reviewed, "approved": approved, "total": len(reviewed)}
